Cython/Compiler/ExprNodes.py

```diff
@@ -6832,160 +6772,154 @@ def generate_unpacking_code(self, code, self_arg, arg_offset_cname, vectorcall_c
-            self.function.free_temps(code)
-        else:
-            code.put_decref_clear(function, py_object_type)
-            code.funcstate.release_temp(function)
-        code.putln("}")
+            space_for_selfarg_var=space_for_selfarg,
+            method_obj=function,
+        )
 
-class PyDirectMethodCallNode(PyMethodCallNode):
-    """
-    Skip the getattr stage and use PyObject_VectorCallMethod
+        function_caller = self.select_utility_code(code)
 
-    function_obj   ExprNode
-    """
+        # Actually call the function.
 
-    def __init__(self, pos, **kwds):
-        super().__init__(pos, **kwds)
-        self.function_obj = self.function.obj
+        code.putln("{")
+        extra_keyword_args = f"+ ((CYTHON_VECTORCALL) ? {len(kwargs_key_value_pairs)} : 0)" if kwargs_key_value_pairs else ""
 
-    def select_utility_code(self, kwargs_key_value_pairs, code):
-        # returns the function cname
-        if kwargs_key_value_pairs:
-            code.globalstate.use_utility_code(
-                UtilityCode.load_cached("PyObjectVectorCallMethodKwBuilder", "ObjectHandling.c"))
-            return "__Pyx_Object_VectorcallMethod_CallFromBuilder"
-        else:
-            code.globalstate.use_utility_code(
-                UtilityCode.load_cached("PyObjectFastCallMethod", "ObjectHandling.c"))
-            return "__Pyx_PyObject_FastCallMethod"
+        # To avoid passing an out-of-bounds argument pointer in the no-args case,
+        # we need at least two entries, so we pad with NULL and point to that.
+        # See https://github.com/cython/cython/issues/5668
+        args_list = ', '.join(arg.py_result() for arg in args) if args else "NULL"
+        code.putln(
+            f"PyObject *{Naming.callargs_cname}[{(len(args) + 1) if args else 2:d}{extra_keyword_args}] = {{{self_arg}, {args_list}}};"
+        )
 
-    def generate_evaluation_code(self, code):
-        code.mark_pos(self.pos)
-        self.allocate_temp_result(code)
+        keyword_variable = ""
+        keyword_nodes_to_clean_up = kwargs_key_value_pairs
 
-        function_name = code.get_py_string_const(self.function.attribute)
+        if kwargs_key_value_pairs:
+            keyword_variable = code.funcstate.allocate_temp(py_object_type, manage_ref=True)
+            self.generate_keyvalue_args(code, args, kwargs_key_value_pairs, keyword_variable)
+        elif self.kwdict:
+            keyword_variable = self.kwdict.result()
+            keyword_nodes_to_clean_up = [self.kwdict]
 
-        args = self.arg_tuple.args
-        kwargs_key_value_pairs = self.kwdict.key_value_pairs if isinstance(self.kwdict, DictNode) else None
-        self.generate_evaluate_args(code)
+        code.putln(
+            f"{self.result()} = {function_caller}("
+            f"{function}, "
+            f"{Naming.callargs_cname}+{space_for_selfarg}, "
+            f"({len(args)+1:d}-{space_for_selfarg}) | ({space_for_selfarg}*__Pyx_PY_VECTORCALL_ARGUMENTS_OFFSET)"
```

Comment:
Small detail that I think you've lost here - it's worth setting `__Pyx_PY_VECTORCALL_ARGUMENTS_OFFSET` for the use_method variant (although it probably only matters in the case where the method call optimization fails)

Reply:
Are you sure? The way I read it, we either pass `self` as first argument or prepend it to the array and set the flag. Which case is missing?

Reply:
For VectorcallMethod specifically

> `nargsf` is the number of positional arguments including `args[0]`, plus [PY_VECTORCALL_ARGUMENTS_OFFSET](https://docs.python.org/3/c-api/call.html#c.PY_VECTORCALL_ARGUMENTS_OFFSET) if the value of `args[0]` may temporarily be changed.

So in this case it's saying that we don't care if they mess with `args[0]` (i.e. the array doesn't come from a tuple that's shared elsewhere, or similar).

It makes no difference if it's actually a method as we hope; it makes a small difference if it isn't a method because it lets it avoid reallocating the array.